v2/car_reader_seeker.go

```diff
@@ -0,0 +1,107 @@
+package car
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"golang.org/x/xerrors"
+)
+
+// CarReaderSeeker wraps CarOffsetWriter with a ReadSeeker implementation.
+// Note that Read and Seek are not thread-safe, they must not be called
+// concurrently.
+type CarReaderSeeker struct {
+	parentCtx context.Context
+	size      uint64
+	offset    int64
+
+	cow           *CarOffsetWriter // 🐮
+	reader        *io.PipeReader
+	writeCancel   context.CancelFunc
+	writeComplete chan struct{}
+}
+
+var _ io.ReadSeeker = (*CarReaderSeeker)(nil)
+
+func NewCarReaderSeeker(ctx context.Context, payloadCid cid.Cid, bstore blockstore.Blockstore, size uint64) *CarReaderSeeker {
+	cow := NewCarOffsetWriter(payloadCid, bstore)
+
+	return &CarReaderSeeker{
+		parentCtx:     ctx,
+		size:          size,
+		cow:           cow,
+		writeComplete: make(chan struct{}, 1),
+	}
+}
+
+// Note: not threadsafe
+func (c *CarReaderSeeker) Read(p []byte) (n int, err error) {
+	if uint64(c.offset) >= c.size {
+		return 0, fmt.Errorf("cannot read from offset %d >= file size %d", c.offset, c.size)
+	}
+
+	// Check if there's already a write in progress
+	if c.reader == nil {
+		// No write in progress, start a new write from the current offset
+		// in a go routine
+		writeCtx, writeCancel := context.WithCancel(c.parentCtx)
+		c.writeCancel = writeCancel
+		pr, pw := io.Pipe()
+		c.reader = pr
+
+		go func() {
+			err := c.cow.Write(writeCtx, pw, uint64(c.offset))
+			if err != nil && !xerrors.Is(err, context.Canceled) {
+				pw.CloseWithError(err)
```

Comment:
This is a little convoluted but:
- Read and Seek should not be called concurrently
- The context is only cancelled by Seek
- Therefore any reads will have finished when the context is cancelled
- Therefore there should not be a read in progress anyway. We just want to discard any bytes that have been written.

Reply:
It's not necessary that the context is only cancelled by Seek. You also pass in a parent context in there which would be cancelled by the creator of `ReadSeeker`. Therefore, context cancellation errors on write should be propagated to the reader.

Reply:
^ I think I agree with this, you're accepting a parent context and it's making its way down here, the cancellation could be for any number of reasons very far from this API, so having that bubble back up would seem to be the appropriate behaviour

Reply:
Oh... actually now I'm looking at `Seek()` it's not that simple is it? we're explicitly using the cancel on the new context created for this in the case of a seek that comes in while we're writing. So the cancellation could come from internal, in `Seek()` _or_ external from `parentCtx`.

But, the cancellation from `Seek()` shouldn't ever show up from the `CloseWithError()` here since the reader is discarded there. So it shouldn't matter if we pass it on since there's no check for it. But since it is passed on with a standard `Read()`, then the `parentCtx` cancellation error will propagate, but will currently be missed.

So it's still correct to pass the error on regardless of where it comes from; as long as we keep it clear in the API docs that this isn't threadsafe (which you've done, but we need to retain when we add more docs here).